### ISSUE TITLE
New savegame system

### ISSUE THREAD

**Issue by Xenega** [singularity/singularity#145]
Related: #52 #56 #122 
PR: #123 #144

Ok, let rewrite the savegame system.

1/ Separate mutable and immutable data.

Locations, events and techs use the same class for its mutable and immutable data. We must separate it the same way others class separate it, i.e. with XClass.

TODO: 
- [x] Migrate data to LocationClass, EventClass and TechClass
- [ ] Add Location, Event and Tech list to Player

1-bis/ (Optionnal) Rename XClass and self.type.

This is ill-named and create confusion. I would personally go for XSpec and self.spec. 

2/ Update savegame with new data.

If 1 was correctly executed, this will be simple. We need to update list in Player. However, the complex stuff happen when we remove things.  Two solutions :
- Nice way: We create rules and special data to handle the suppression.
- Brutal way: We remove everything. If the user have no base, we recreate a University Computer. 

TODO: 
- [ ] Update Player the nice way or brutal way.

2-bis/ (Optionnal) Remove computed value from savegame
 
We should not save everything but real mutable state. Everything that could be computed should not be included. We should do the inverse of the current state which is modified after each trigger, and instead fully re-compute the state after each trigger. It allows to take into account every change made after the savegame.

TODO: 
- [ ] Remove computed value from savegame.
- [ ] Re-compute everything that can be (bonus, ...).

3/ Correct the old savegame.

The old savegame need to be correctly reloaded. A problem found is that pickle use string where we use now unicode. Others problems could be hidden.

TODO
- [ ] Replace string with unicode.

4/ Create the new savegame

Basically, we need two things:
- A small header to contains metadata.
- The savegame data with its format.

I wanted personally to use String + Json with GZIP. The metadata need to be parsed quickly, so string is good and we can limit the number of characters allowed. Json have the advantage to have no dependencies needed. GZIP allows to reduce the size of the savegame and are sufficiently fast.

TODO
- [ ] New format: String + Json with GZIP
- [ ] New metadata: Date and difficulty.

Conclusion:
The problems with the current savegame is multiple and difficult to solve with few commit. So, I thought to add a new branch to handle all of it and allows other to work on it.

**Comment by nthykier** [comment]
For 2/, I think the correct answer is to do it on a case-by-case basis.  When we can (reasonably) convert to something comparable, we should.  When we cannot, we should fallback to discarding the data.

Another aspect we should consider is how we save the (mutable) state.  At the moment, we store all values as they are.  That makes it "simpler" but it also means we cannot easily adjust the bonuses already applied in a save game (e.g. if we change a tech to give a different bonus adjustment).  My take on that is that we should (eventually) recompute the bonuses again by re-applying the bonus one by one (with the new game data).

That brings us to, what state do we need to store?

 * Buyables in general: 
   * What kind of buyable is it? (e.g. `tech`, `item`)
   * What is the id of the buyable? (e.g. `Simulacra`, `Quantum Computer`)
   * If it complete (bool)?
   * How many resources have been invested in it?
   * Their count (when different from 1)
   * For Bases: Items, location, power-state, started_at, name, suspicion
 * Groups:
  * Their id (e.g. "news")
  * Their current suspicion
 * Player:
   * current time
   * Cash (plus partial cash)
   * The log
   * difficulty level
   * last_discovery + prev_discovery
   * current CPU allocation (cpu_pool + relevant bits of cpu_usage)
 * Events:
   * The event id
   * For unique events: Whether they have triggered already
   * For active events: When they will expire if at all (we do not track this at the moment; currently all events are permanent and non-unique events can trigger indefinitely many times stacking to an infinite bonus/penalty).

**Comment by Xenega** [comment]
> Another aspect we should consider is how we save the (mutable) state. At the moment, we store all values as they are. That makes it "simpler" but it also means we cannot easily adjust the bonuses already applied in a save game (e.g. if we change a tech to give a different bonus adjustment). My take on that is that we should (eventually) recompute the bonuses again by re-applying the bonus one by one (with the new game data).

Yes, absolutely.

> That brings us to, what state do we need to store?

One solution could to differentiate someways between computed value and mutable state.

**Comment by nthykier** [comment]
@Xenega: I am not sure I understand the need for moving events, techs and locations into the player object.  Can you expand a bit on the rationale behind that or what you are hoping will be solved by that?

**Comment by Xenega** [comment]
@nthykier The rationnal is that we separate what is game data and what is game state. Global list become LocationClass, TechClass and EventClass list. So we need to save now Location, Tech and Event. So we just put them in player.

**Comment by nthykier** [comment]
I agree that we want to separate game state vs. game data and I like the idea of the global dictionaries being static game data rather than state.  However, I think shoving everything into the `Player` object will just turn it into an God-object/Monster-object, which becomes a long term maintenance burden. Note, I am open to fixing the savegame issue first and deal with the monster-object issue afterwards; I just do not want us to be blinded by the savegame issue that we overlook that we might be introducing a separate issue elsewhere as a part of the fix.

**Comment by Xenega** [comment]
The goal is to make `Player` the root of all game data. But yes, there always the danger to make it a god object. `Player` already have too many responsibility. We can separate it with new class after.
